.changeset/mean-horses-kiss.md

```diff
@@ -0,0 +1,26 @@
+---
+'astro': minor
+---
+
+Adds a new `injectTypes()` utility to the Integration API and refactors how type generation works
+
+Integrations can now use a new `injectTypes()` utility in the `astro:config:done` hook. This utility allows an integration to inject types into a user's project. `filename` must end with `".d.ts"` and `content` must be valid TypeScript (it will be formatted).
+
+Under the hood, it will create a file at `/.astro/integrations/<normalized_integration_name>/<normalized_filename>.d.ts` and create references to it. `injectTypes()` returns a URL to the normalized path.
+
+```js
+const path = injectTypes({
```

Comment:
What's the point of it returning the path?

Reply:
See what we're doing in core for content sync (ie. overriding the dts file with content updates)? I can imagine some people wanting to do this in integrations. Since we normalize the filename, people can't easily construct the path themselves so we give it to them